### PR TITLE
Feature: Add CVE Search, Filtering, Caching, and Autocomplete

### LINKED ISSUE #5050: Enhanced CVE Integration
## Overview
This issue tracks the enhancement of CVE (Common Vulnerabilities and Exposures) integration in OWASP BLT. Currently, basic CVE functionality exists (CVE ID input, score fetching from NVD API, and display), but we can significantly improve it to provide more value to users and organizations.

## Current State
- `cve_id` and `cve_score` fields in Issue model
- NVD API integration via `get_cve_score()` method
- CVE ID validation in report form
- CVE score display on issue detail pages
- CVE-based analytics in organization dashboards

### 1. Enhanced CVE Data Model
- Create dedicated `CVE` model to store comprehensive CVE data
- Store description, published date, affected products, references
- Support CVSS v2 and v3 metrics
- Link to Issue model while maintaining backward compatibility

### 2. Improved NVD API Integration
- Refactor to fetch full CVE details instead of just base score
- Better error handling and rate limit management
- Support for CVSS v3.1 metrics

### 3. Caching Layer
- Implement caching for CVE data to reduce API calls
- Cache invalidation strategy
- Handle cache misses gracefully

### 4. CVE Search & Filtering
- Search/filter issues by CVE ID
- Filter by CVE score range
- Database indexes for performance

### 5. API Enhancements
- Dedicated CVE endpoints (`/api/v1/cve/`)
- Enhanced IssueSerializer with richer CVE data
- CVE statistics endpoint

### 6. UI/UX Improvements
- Real-time CVE ID autocomplete in report form
- Enhanced CVE display with severity badges
- Expandable CVE details section
- Visual indicators for CVE severity

### 7. Advanced Features (Optional)
- Background tasks for periodic CVE score updates
- CVE history tracking
- CVE-based notifications/alerts

## Implementation Plan

This will be implemented incrementally across multiple PRs:

1. Enhanced CVE data model
2. Improved NVD API integration
3. Caching layer
4. CVE search & filtering (backend)
5. CVE API endpoints
6. CVE autocomplete (frontend)
7. Enhanced CVE display (UI)
8. Advanced features (optional)

## Benefits
- Better user experience with search and autocomplete
- Improved performance through caching
- Richer CVE information for better decision-making
- Enhanced analytics and reporting capabilities
- Better scalability with background tasks

## Technical Stack
- Backend: Django models, views, API endpoints
- Caching: Django cache framework
- API: NVD API v2.0 integration
- Frontend: Tailwind CSS (following project standards)

## Related
- Existing CVE fields: `website/models.py` (lines 580-581)
- Current implementation: `website/models.py` (lines 627-641)
- CVE display: `website/templates/includes/issue_right_info.html`

## Acceptance Criteria
- [ ] Enhanced CVE data model created
- [ ] Full CVE details fetched from NVD API
- [x] Caching implemented
- [ ] Search and filtering functional
- [ ] API endpoints created
- [ ] UI improvements implemented
- [x] Backward compatibility maintained
- [x] Tests written and passing

### PR DESCRIPTION
**Pull request by Pritz395**
## summary
This PR implements a comprehensive CVE (Common Vulnerabilities and Exposures) integration system for OWASP BLT, enabling users to search, filter, and manage issues by CVE ID and CVSS scores. The implementation includes a robust caching layer to minimize NVD API calls, efficient database indexing, and a user-friendly autocomplete feature.

### 1. No CVE Search Functionality
**Problem:** Users cannot search for issues by CVE ID in the web interface, making it difficult to find related vulnerabilities.
**Solution:** Added `cve:` prefix search support in the web search interface (`search_issues` function) with case-insensitive matching and proper visibility filtering.

### 2. No CVE Filtering in API
**Problem:** The REST API lacks filtering capabilities for CVE ID and CVSS scores, preventing programmatic access to CVE-specific issues.
**Solution:** Implemented API filtering with `cve_id`, `cve_score_min`, and `cve_score_max` query parameters, including validation and range checking.

### 3. Excessive NVD API Calls
**Problem:** The existing `get_cve_score()` method makes direct API calls to NVD for every request, leading to rate limiting and performance issues.
**Solution:** Implemented a comprehensive caching layer (`website/cache/cve_cache.py`) with:
- 24-hour cache timeout
- Cache-backed locking to prevent duplicate API calls
- Exponential backoff for 429 rate limit errors
- Sentinel values to cache `None` results
- Double-check pattern for race condition prevention

### 4. No CVE Autocomplete
**Problem:** Users must manually type full CVE IDs when reporting issues, leading to typos and inconsistent formatting.
**Solution:** Added autocomplete functionality with:
- Backend API endpoint (`/api/v1/cve/autocomplete/`)
- Frontend JavaScript with keyboard navigation (Arrow keys, Enter, Escape)
- XSS prevention using `textContent` instead of `innerHTML`
- Visibility filtering to exclude hidden/hunt issues

### 5. Database Performance Issues
**Problem:** CVE-related queries are slow due to lack of indexes on `cve_id` and `cve_score` fields.
**Solution:** Added database indexes (`issue_cve_id_idx`, `issue_cve_score_idx`) to optimize filtering and search queries.

### 6. Database Schema Limitations
**Problem:** 
- `cve_score` field cannot store scores of 10.0 (max_digits=2 allows only 0.0-9.9)
- `cve_id` field max_length=16 may be insufficient for longer CVE IDs
**Solution:** 
- Increased `cve_score.max_digits` from 2 to 3
- Increased `cve_id.max_length` from 16 to 20
- Created migrations to update existing database schema

### 7. Security Vulnerabilities
**Problem:**
- CVE IDs are not URL-encoded when making API calls, allowing injection attacks
- CVE ID format is not validated at the model level
**Solution:**
- URL encoding using `urllib.parse.quote()` in API calls
- Regex-based validation (`CVE_ID_PATTERN`) in `normalize_cve_id()`
- Model-level validation in `Issue.clean()` method

### 8. Data Quality Issues
**Problem:**
- CVE IDs are stored without normalization, causing case-sensitivity mismatches
- Whitespace-only CVE IDs are stored as-is instead of being rejected
- CVE scores are not populated when issues are created
**Solution:**
- Normalize CVE IDs to uppercase on save in all creation paths
- Use case-insensitive matching (`__iexact`, `__istartswith`) in queries
- Set `cve_id` to `None` when normalization returns empty string
- Populate `cve_score` automatically in all issue creation paths (web form, API, hunt submissions)

### 9. Race Conditions
**Problem:** Concurrent requests can cause duplicate API calls and inconsistent data updates.
**Solution:**
- Cache-backed locking with timeout handling
- `select_for_update()` in API create endpoint to prevent race conditions
- Proper lock cleanup in finally blocks

### 1. CVE Caching Layer (`website/cache/cve_cache.py`)
- **`get_cached_cve_score(cve_id)`**: Main entry point with cache-first strategy
- **`fetch_cve_score_from_api(cve_id)`**: Direct NVD API integration with retry logic
- **`normalize_cve_id(cve_id)`**: CVE ID normalization and validation
- **Cache lock mechanism**: Prevents duplicate API calls under concurrent load
- **Exponential backoff**: Handles 429 rate limit errors gracefully
- **CVSS version preference**: Prefers v3.1 → v3.0 → v2.0

### 2. CVE Search (`website/views/issue.py`)
- Web search with `cve:` prefix (e.g., `cve:CVE-2024-1234`)
- Case-insensitive matching
- Respects visibility rules (hidden issues, hunt issues)
- Returns results ordered by creation date

### 3. CVE API Filtering (`website/api/views.py`)
- **`cve_id` filter**: Exact match (case-insensitive)
- **`cve_score_min` filter**: Minimum CVSS score (0-10)
- **`cve_score_max` filter**: Maximum CVSS score (0-10)
- **Range validation**: Ensures `min_score <= max_score`
- **Input validation**: Handles invalid inputs gracefully with logging
- **Automatic CVE normalization**: Normalizes and populates scores on issue creation

### 4. CVE Autocomplete (`website/views/issue.py`, `website/templates/report.html`)
- Backend API endpoint with visibility filtering
- Frontend JavaScript with debouncing (300ms)
- Keyboard navigation (ArrowUp, ArrowDown, Enter, Escape)
- Click-to-select functionality
- XSS prevention using DOM manipulation
- Orders suggestions by most recent usage

### 5. Database Migrations
- **0257_add_cve_indexes.py**: Adds indexes on `cve_id` and `cve_score`
- **0258_fix_cve_score_max_digits.py**: Increases `max_digits` to 3
- **0259_increase_cve_id_max_length.py**: Increases `max_length` to 20

### 6. Model Enhancements (`website/models.py`)
- **`clean()` method**: Model-level CVE ID format validation
- **`get_cve_score()` refactor**: Delegates to caching layer
- **Database indexes**: Added to `Meta.indexes`

### Caching Strategy
- **Cache timeout**: 24 hours (configurable via `CVE_CACHE_TIMEOUT`)
- **Lock timeout**: 30 seconds (configurable via `CVE_CACHE_LOCK_TIMEOUT`)
- **Lock wait timeout**: 5 seconds (configurable via `CVE_CACHE_LOCK_WAIT_TIMEOUT`)
- **Sentinel value**: `CACHE_NONE` to distinguish cached `None` from cache miss

### API Rate Limiting
- **Max retries**: 3 (configurable via `CVE_API_MAX_RETRIES`)
- **Exponential backoff**: Base delay 0.5s, doubles on each retry
- **Timeout**: 10 seconds per request (configurable via `CVE_API_TIMEOUT`)

### Security Measures
- URL encoding for CVE IDs in API calls
- Regex validation: `^CVE-\d{4}-\d{4,7}$`
- Model-level validation in `Issue.clean()`
- XSS prevention in autocomplete (no `innerHTML`)

### Performance Optimizations
- Database indexes on `cve_id` and `cve_score`
- Case-insensitive queries to handle unnormalized data
- Cache-first strategy to minimize API calls
- Lock mechanism to prevent duplicate fetches

### Test Coverage
- **`website/tests/test_cve_cache.py`**: 497 lines, comprehensive caching tests
  - CVE ID normalization (valid, invalid, empty)
  - Cache hit/miss scenarios
  - Lock acquisition and timeout handling
  - API retry logic and exponential backoff
  - Error handling (timeout, connection, HTTP errors)
  - Sentinel value handling

- **`website/tests/test_cve_filtering.py`**: 304 lines, filtering and search tests
  - API filtering by `cve_id`, `cve_score_min`, `cve_score_max`
  - Range validation (min > max)
  - Invalid input handling
  - Web search with `cve:` prefix
  - Visibility filtering (hidden issues, hunt issues)
  - Case-insensitive matching
  - Result ordering

### Test Configuration
- Added `pytest.ini` for Django test configuration
- All tests use pytest fixtures and Django test client

### New Files
- `website/cache/__init__.py`: Package initializer
- `website/cache/cve_cache.py`: CVE caching implementation (295 lines)
- `website/migrations/0257_add_cve_indexes.py`: Database indexes
- `website/migrations/0258_fix_cve_score_max_digits.py`: Schema fix
- `website/migrations/0259_increase_cve_id_max_length.py`: Schema fix
- `website/tests/test_cve_cache.py`: Caching tests (497 lines)
- `website/tests/test_cve_filtering.py`: Filtering tests (304 lines)
- `pytest.ini`: Pytest configuration

### Modified Files
- `website/models.py`: Model updates, validation, indexes
- `website/api/views.py`: API filtering and CVE normalization
- `website/views/issue.py`: Search, autocomplete, normalization helper
- `website/templates/report.html`: Autocomplete UI and JavaScript
- `blt/urls.py`: Autocomplete endpoint route

## Related Issues
Fixes #5050

## Breaking Changes
None. All changes are backward compatible. Existing CVE data will be normalized on next update.

## Migration Notes
1. Run migrations: `python manage.py migrate`
2. Existing CVE IDs will be normalized on next issue update
3. Cache will be populated on first CVE score fetch

## Configuration
All CVE-related settings are optional and have sensible defaults:
- `CVE_CACHE_TIMEOUT`: Cache timeout in seconds (default: 86400)
- `CVE_API_TIMEOUT`: API request timeout (default: 10)
- `CVE_API_MAX_RETRIES`: Max retry attempts (default: 3)
- `CVE_RATE_LIMIT_BACKOFF_BASE`: Exponential backoff base (default: 0.5)
- `CVE_CACHE_LOCK_TIMEOUT`: Lock timeout (default: 30)
- `CVE_CACHE_LOCK_WAIT_TIMEOUT`: Lock wait timeout (default: 5)
- `CVE_CACHE_LOCK_WAIT_INTERVAL`: Lock wait interval (default: 0.2)

### Updates after maintainer feedback

- Reverted all non-CVE changes (GitHub webhooks, hackathon leaderboard, user views, utils, etc.) so this PR now only touches:
  - CVE search & filtering
  - CVE caching
  - CVE autocomplete
  - CVE-related migrations
  - CVE tests
- Fixed a critical transaction bug by wrapping `select_for_update()` in `transaction.atomic()`.
- Addressed CodeRabbit’s BLE001 warning by removing redundant broad exception handling.
- Added `pytest` / `pytest-django` to dev dependencies so the CVE test suite can run.
- All CVE tests pass:
  - `website/tests/test_cve_cache.py`
  - `website/tests/test_cve_filtering.py`
- One failing test remains in `website/tests/test_core.py::StatusPageTests::test_status_page_has_required_context`, which is pre-existing and unrelated to CVE. 

## Visual

1. Reporting a valid bug "CVE-2023-44487"
Screenshot 2026-02-02 at 10 55 42 PM

2. Filtering by CVE ID 

Screenshot 2026-02-02 at 11 00 20 PM

3. Filtering by CVE Score Range
Screenshot 2026-02-02 at 11 01 38 PM

4. CVE Autocomplete
Screenshot 2026-02-02 at 11 02 49 PM

## Summary by CodeRabbit

* **New Features**
  * CVE score caching and automated CVE score population for issues.
  * CVE autocomplete endpoint and client autocomplete on the report form.
  * Issue filtering by CVE ID and by CVE score range.

* **Bug Fixes**
  * Improved CVE ID validation, normalization, and user-facing error messages.
  * Safer CVE score updates to avoid long-running transactions and reduce locking.

* **Tests**
  * Extensive tests covering CVE caching, fetching, filtering, autocomplete, and error scenarios.

### CODE REVIEW

**Review by DonnieBLT — CHANGES_REQUESTED** [review]
please combine all CVE related changes into one PR and undo any changes not related to CVE

**Comment by Pritz395** [comment]
@coderabbitai I've cleaned up the commit history for PR #5057. The PR now has:
✅ Clean Commit History:
Reduced from 39 messy commits to 3 clean, logical commits
All commits are CVE-related only
No merge commits with non-CVE changes
✅ Final Diff:
Only 12 files changed, all CVE-related:
website/cache/ - CVE caching module
website/models.py - CVE field updates and caching integration
website/api/views.py - CVE filtering
website/views/issue.py - CVE search & autocomplete
website/templates/report.html - CVE autocomplete UI
website/migrations/ - Database indexes and schema fixes
website/tests/ - Comprehensive test coverage
blt/urls.py - CVE autocomplete endpoint
pytest.ini - Test configuration
✅ All Issues Fixed:
Database schema: cve_score max_digits fixed (2→3 for CVSS 10.0)
Data population: cve_score populated in all creation paths
Normalization: CVE IDs normalized on save, case-insensitive matching in queries
No non-CVE changes in commit history or final diff
Can you please verify:
The final diff (origin/main to HEAD) contains only CVE-related changes
The commit history is clean (3 commits, all CVE-related)
No non-CVE files are present in the PR

**Review by DonnieBLT — CHANGES_REQUESTED** [review]

**Comment by Pritz395** [comment]
@DonnieBLT , I've addressed the review comments! Thanks

**Comment by Pritz395** [comment]
I'll fix this. Thanks

**Comment by Pritz395** [comment]
@coderabbitai what are the conflicts occuring with the base branch?

**Comment by Pritz395** [comment]
@coderabbitai what are the pending issues to be resolved? cuz the test runs are failing due to migration checks which isn't a code issue

**Comment by Pritz395** [comment]
@DonnieBLT PR should be error free now.

**Comment by Pritz395** [comment]
@coderabbitai what are the conflicts that must be resolved?

**Review by Jayant2908 — CHANGES_REQUESTED** [review]
Great Work! Could you ensure these few things are dealt with!?
1)CVE search datasets are large; missing indexes will cause slow queries / DB load.
Ensure indexes exist on commonly filtered fields (e.g., cve_id, published_date, vendor/product, score).I f using full-text search: confirm GIN/TSVector indexes. Check for N+1 queries in list/detail rendering.

2)Direct calls to external CVE/NVD APIs without backoff or throttling can hit provider rate limits and degrade the app.
Issues to address:

- Ensure API client supports retries with exponential backoff on 429 and 5xx.

- Long-running fetches should be moved to background tasks instead of blocking web requests.

- Scheduled sync jobs should batch external requests and respect provider rate limits (sleep / throttling).

3)Expensive or unsafe queries can cause SQL injection-like behavior, slow queries, or DoS.
Issues to address:

- Validate and normalize all incoming query parameters (max length, allowed characters).

- Confirm ORM parameterization is used everywhere—no raw query string concatenation.

- Pagination must have strict maximums.

- Autocomplete-like endpoints should enforce strict limits.

4)Caching API failure responses (5xx) can freeze the UI with stale errors.
Issues to address:

- Only cache successful responses.

- If error responses are cached, TTL must be short and use separate cache keys.

**Comment by Pritz395** [comment]
@coderabbitai why did the run-tests fail and does the fix lie within PR scope

**Review by Jayant2908 — APPROVED** [review]
LGTM now!! Just ensure that tasks like full cache warming or history refreshes are offloaded to background workers and show last-updated timestamps for CVE data to users to see data freshness.

**Review by arnavkirti — APPROVED** [review]
LGTM. Great Work!!!

**Review by Jayant2908 — CHANGES_REQUESTED** [review]
Hey Preetham Can you check on the 2nd issue mentioned by coderabbitai!?

**Comment by Pritz395** [comment]
@coderabbitai give the fix for issue 2.

File: website/api/views.py [hunk 1]
```diff
@@ -255,10 +309,64 @@
...
             return Response({"error": "Max limit of 5 images!"}, status=status.HTTP_400_BAD_REQUEST)
 
-        data = super().create(request, *args, **kwargs).data
+        # Wrap super().create() to catch model-level ValidationError from Issue.save()
+        try:
+            data = super().create(request, *args, **kwargs).data
+        except ValidationError as e:
+            # Convert model-level ValidationError to HTTP 400 Bad Request
+            error_message = e.message if hasattr(e, "message") else str(e)
+            return Response({"error": f"Invalid CVE ID: {error_message}"}, status=status.HTTP_400_BAD_REQUEST)
```
[the remark this answers is not in the document]
Comment (Pritz395) [reply website/api/views.py]: fixed!

File: website/api/views.py [hunk 2]
```diff
@@ -255,10 +309,73 @@
...
 
-        data = super().create(request, *args, **kwargs).data
+        # Wrap super().create() to catch model-level ValidationError from Issue.save()
+        try:
+            data = super().create(request, *args, **kwargs).data
+        except ValidationError as e:
+            # Convert model-level ValidationError to HTTP 400 Bad Request
+            error_message = e.message if hasattr(e, 'message') else str(e)
+            return Response(
+                {"error": f"Invalid CVE ID: {error_message}"},
```
[the remark this answers is not in the document]
Comment (Pritz395) [reply website/api/views.py]: fixed

**Comment by Pritz395** [comment]
@Jayant2908 feel free to look into this again!

**Review by Jayant2908 — APPROVED** [review]
LGTM Now! Nice work!

**Review by Nachiket-Roy — APPROVED** [review]
A couple of small duplications here and there, no major issues

**Comment by Pritz395** [comment]
@coderabbitai what's causing the run-test failure? also is there anything else pending to be addressed?

**Review by Jayant2908 — APPROVED** [review]
LGTM! Nice work addressing to the issues.

**Comment by Pritz395** [comment]
@coderabbitai why does it still show needs peer-review though it's been approved by multiple contributors?

**Comment by Pritz395** [comment]
@coderabbitai so I just confirmed with the maintainer. He basically requested that the tests be moved into a separate follow-up branch/PR and not part of this PR. So what's the best way to do it? Give solution.

**Comment by Pritz395** [comment]
@coderabbitai verify if this PR is now test-free and as per maintainer's request

**Review by Jayant2908 — CHANGES_REQUESTED** [review]
Most of the things look good, Do think of adding a rate-limiter is what I would suggest.

**Review by Nachiket-Roy — COMMENTED** [review]
I think it’s worth adding the limiter now. As said by CR even if traffic is low one curious user or a basic script can send dozens of requests per second.

**Review by Jayant2908 — COMMENTED** [review]
I agree with @Nachiket-Roy.

**Review by Nachiket-Roy — CHANGES_REQUESTED** [review]
I think most of these can be deferred to a follow-up. Could you also attach a quick demo showing how this works?

**Review by Jayant2908 — APPROVED** [review]
LGTM!

**Comment by Pritz395** [comment]
@DonnieBLT could you please take a look at this PR

**Review by DonnieBLT — CHANGES_REQUESTED** [review]
Please attach a demo

**Review by Nachiket-Roy — APPROVED** [review]
Thanks! A short video would be even better, but this works too.

**Review by yashhzd — COMMENTED** [review]
Solid CVE integration work. Reviewed the caching layer, API filtering, autocomplete, model changes, and search enhancements. Here's my feedback:

## What works well

1. **Cache architecture** — The `get_cached_cve_score()` flow is well-designed: lock tokenization with UUIDs prevents lock stomping, sentinel values distinguish "CVE not found" from cache miss, and the `had_error` flag correctly skips caching on API/network errors so stale error states don't persist.

2. **NVD API handling** — Exponential backoff with `Retry-After` header support, CVSS version preference (v3.1 → v3.0 → v2.0), URL encoding via `quote(cve_id, safe="")`, and configurable timeouts. This matches NVD's recommended integration patterns.

3. **XSS prevention in autocomplete** — Using `textContent` instead of `innerHTML` for suggestion rendering. ARIA attributes (`role="combobox"`, `aria-expanded`, `role="option"`) for accessibility. Stale response detection via `lastQuery` and `requestId`.

4. **Model-level validation** — CVE normalization in both `clean()` and `save()` ensures consistency regardless of entry path (admin, API, form). The `update_fields` handling in `save()` to include `cve_id` when it gets normalized is a subtle but important detail.

5. **Test coverage** — 1198 lines of tests covering cache hit/miss, lock contention, API retry logic, score range filtering, autocomplete visibility rules, and edge cases.

## Potential concerns

1. **`_release_cache_lock` has a TOCTOU race** — The `cache.get(lock_key)` → compare → `cache.delete(lock_key)` pattern isn't atomic. Between the get and delete, another worker could acquire the expired lock, and then the delete would release *their* lock. For Redis backends, a Lua script (`if redis.call("get",KEYS[1]) == ARGV[1] then return redis.call("del",KEYS[1])`) would make this safe. For LocMemCache (single-process) this is fine, but worth documenting which cache backend is assumed.

2. **Synchronous `time.sleep()` blocks Django workers** — Both `_wait_for_cache_fill()` and `fetch_cve_score_from_api()` use `time.sleep()`. Under concurrent load with many unique CVE lookups, sleeping workers could exhaust the WSGI pool. Consider documenting this as a known limitation, or deferring CVE score population to a Celery task so it never blocks request handling.

3. **Search safe-chars regex compiled per request** — In `search_issues()`, `re.match(r"^[a-zA-Z0-9\s\-\.:_/@\?&=\%\#\+]+$", query)` compiles a new regex on every search. Minor performance hit — could be a module-level `re.compile()` constant.

4. **`cve_autocomplete` has no max query length check** — Minimum (3 chars) is validated, but there's no upper bound. A very long query string would be passed directly to `cve_id__istartswith`, which could generate an expensive LIKE query. Consider capping at ~30 characters (longest valid CVE ID is ~16 chars).

5. **Inline `import re` in `search_issues`** — Line ~569 has `import re` inside the function body, but `re` is already imported at the module level in several other files. This should use the module-level import (it's also imported at the top of `cve_cache.py`).

Overall this is a high-quality, well-tested addition. The caching layer is particularly well-thought-out. Nice work.

**Review by DonnieBLT — APPROVED** [review]
